src/agent/application.cpp

```diff
@@ -60,45 +60,50 @@ Application::Application(const std::string               &aInterfaceName,
 #else
     , mBackboneInterfaceName(aBackboneInterfaceNames.empty() ? "" : aBackboneInterfaceNames.front())
 #endif
-    , mHost(mInterfaceName.c_str(), aRadioUrls, mBackboneInterfaceName, /* aDryRun */ false, aEnableAutoAttach)
 #if OTBR_ENABLE_MDNS
     , mPublisher(Mdns::Publisher::Create([this](Mdns::Publisher::State aState) { this->HandleMdnsState(aState); }))
 #endif
+#if OTBR_ENABLE_VENDOR_SERVER
+    , mVendorServer(vendor::VendorServer::newInstance(*this))
+#endif
+{
+    mHost = std::unique_ptr<Ncp::RcpHost>(new Ncp::RcpHost(mInterfaceName.c_str(), aRadioUrls, mBackboneInterfaceName,
+                                                           /* aDryRun */ false, aEnableAutoAttach));
+
 #if OTBR_ENABLE_BORDER_AGENT
-    , mBorderAgent(mHost, *mPublisher)
+    mBorderAgent = std::unique_ptr<BorderAgent>(new BorderAgent(*mHost, *mPublisher));
```

Comment:
I'm still inclined to reserving the initialization list for the constructor.

Reply:
I agree initialization list is better. As discussed offline, let's keep this as a “transition". Because currently all these modules don't work under NCP mode, let's not initialize them and pass the `host` to it at this moment. When later we add NCP support to these modules, we can put them back to initialization list.